File: binary_search/34.py
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        
        if len(nums) == 0:
            return [-1,-1]
        
        # find target index l 
        l, r = 0, len(nums)-1
        while l < r:
            m = l + (r-l)//2
            if nums[m] < target:
                l = m+1
            else:
                r = m
        
        if nums[l] != target:
            return [-1,-1]
        
        index1, index2 = l, l
        while index1 >= 0 and nums[index1] == target:
            index1 -= 1
        while index2 <= len(nums)-1 and nums[index2] == target:
            index2 += 1
        return [index1+1, index2-1]

File: binary_search/34.java
class Solution {
    public int[] searchRange(int[] nums, int target) {
        int[] res = {-1,-1};
        if (nums.length == 0) {
            return res;
        }
        
        int l = 0;
        int r = nums.length-1;
        while (l < r) {
            int m = l + (r-l)/2;
            if (nums[m] < target) {
                l = m + 1;
            } else {
                r = m;
            }
        }
        if (nums[l] != target) {
            return res;
        }
        
        int index1 = l;
        int index2 = l;
        while (index1 >= 0 && nums[index1] == target) {
            index1--;
        }
        while (index2 <= nums.length-1 && nums[index2] == target) {
            index2++;
        }
        res[0] = index1 + 1;
        res[1] = index2 -1;
        return res;
    }
}
